cni/pkg/ebpf/server/redirectionServer.go

```diff
@@ -0,0 +1,592 @@
+	obj := ambient_redirectObjects{}
+	if err := loadAmbient_redirectObjects(&obj, &options); err != nil {
+		return fmt.Errorf("loading objects: %v", err)
+	}
+	r.obj = obj
+	r.ztunnelHostingressFd = uint32(r.obj.ZtunnelHostIngress.FD())
+	ztunnelHostingressInfo, err := r.obj.ZtunnelHostIngress.Info()
+	if err != nil {
+		return fmt.Errorf("unable to load metadata of bfp prog: %v", err)
+	}
+	r.ztunnelHostingressProgName = ztunnelHostingressInfo.Name
+	r.ztunnelIngressFd = uint32(r.obj.ZtunnelIngress.FD())
+	ztunnelIngressInfo, err := r.obj.ZtunnelIngress.Info()
+	if err != nil {
+		return fmt.Errorf("unable to load metadata of bfp prog: %v", err)
+	}
+	r.ztunnelIngressProgName = ztunnelIngressInfo.Name
+
+	r.inboundFd = uint32(r.obj.AppInbound.FD())
+	inboundInfo, err := r.obj.AppInbound.Info()
+	if err != nil {
+		return fmt.Errorf("unable to load metadata of bfp prog: %v", err)
+	}
+	r.inboundProgName = inboundInfo.Name
+	r.outboundFd = uint32(r.obj.AppOutbound.FD())
+	outboundInfo, err := r.obj.AppOutbound.Info()
+	if err != nil {
+		return fmt.Errorf("unable to load metadata of bfp prog: %v", err)
+	}
+	r.outboundProgName = outboundInfo.Name
+	return nil
+}
+
+// Note: this struct should be exactly the same defined in C
+// it will be encoded byte by byte into memory
+type mapInfo struct {
+	Ifindex uint32
+	MacAddr [6]byte
+	Flag    uint8
+	Pad     uint8
+}
+
+func NewRedirectServer() *RedirectServer {
+	if err := checkOrMountBPFFSDefault(); err != nil {
+		log.Fatalf("BPF filesystem mounting on /sys/fs/bpf failed: %v", err)
+	}
+
+	if err := setLimit(); err != nil {
+		log.Fatalf("Setting limit failed: %v", err)
+	}
+
+	r := &RedirectServer{
+		redirectArgsChan: make(chan *RedirectArgs),
+	}
+
+	if err := r.initBpfObjects(); err != nil {
+		log.Fatalf("Init bpf objects failed: %v", err)
+	}
+
+	return r
+}
+
+func checkOrMountBPFFSDefault() error {
+	var err error
+
+	_, err = os.Stat(MapsRoot)
+	if err != nil {
+		if os.IsNotExist(err) {
+			if err := os.MkdirAll(MapsRoot, 0o755); err != nil {
+				return fmt.Errorf("unable to create bpf mount directory: %s", err)
+			}
+		}
+	}
+
+	fst := unix.Statfs_t{}
+	err = unix.Statfs(MapsRoot, &fst)
+	if err != nil {
+		return &os.PathError{Op: "statfs", Path: MapsRoot, Err: err}
+	} else if fst.Type == FilesystemTypeBPFFS {
+		return nil
+	}
+
+	err = unix.Mount(MapsRoot, MapsRoot, "bpf", 0, "")
+	if err != nil {
+		return fmt.Errorf("failed to mount %s: %s", MapsRoot, err)
+	}
+
+	return nil
+}
+
+func setLimit() error {
+	return unix.Setrlimit(unix.RLIMIT_MEMLOCK,
+		&unix.Rlimit{
+			Cur: unix.RLIM_INFINITY,
+			Max: unix.RLIM_INFINITY,
+		})
+}
+
+func (r *RedirectServer) Start(stop <-chan struct{}) error {
+	log.Infof("Starting redirection Server")
+	go func() {
+		for {
+			select {
+			case arg := <-r.redirectArgsChan:
+				if err := r.handleRequest(arg); err != nil {
+					log.Errorf("failed to handle request: %v", err)
+				}
+
+			case <-stop:
+				close(r.redirectArgsChan)
```

Comment:
this is not safe. Any writes to the chan will now panic. We should instead just remove this line

Reply:
fixed